rust/common-graph/src/builder.rs

```diff
@@ -0,0 +1,255 @@
+
+        let edge: Edge = (
+            (node_index.to_owned(), from_port.1.into()),
+            (0, to_port.into()),
+        );
+        self.edges.push(edge);
+        Ok(self)
+    }
+
+    /// Connect one node port to another.
+    pub fn connect<S1, S2, S3, S4>(mut self, from_port: (S1, S2), to_port: (S3, S4)) -> Result<Self>
+    where
+        S1: Into<String>,
+        S2: Into<String>,
+        S3: Into<String>,
+        S4: Into<String>,
+    {
+        let (from_index, to_index) = {
+            let from_name = from_port.0.into();
+            let (_, from_index, _, _) = self.get_node(&from_name)?;
+            let to_name = to_port.0.into();
+            let (_, to_index, _, _) = self.get_node(&to_name)?;
+            (from_index, to_index)
+        };
+
+        let edge: Edge = (
+            (from_index.to_owned(), from_port.1.into()),
+            (to_index.to_owned(), to_port.1.into()),
+        );
+        self.edges.push(edge);
+        Ok(self)
+    }
+
+    fn get_node(&self, label: &str) -> Result<&BuilderNode<T>> {
+        self.nodes
+            .iter()
+            .find(|n| n.0.as_str() == label)
+            .ok_or_else(|| CommonGraphError::GraphBuilderFailure(format!("No node with name {}.", label)))
+    }
+
+    /// Builds a [PortGraph].
+    pub fn build(self) -> Result<PortGraph<T>> {
+        // Note that "inputs" and "outputs" to the graph
+        // are represented by the root node, where graph "inputs"
+        // and "outputs" are node outputs and inputs (reversed).
+        let inputs = self.inputs.unwrap_or_default();
+        let outputs = self.outputs.unwrap_or_default();
+        let root_label = self.root_label.unwrap_or_else(|| "Root".into());
+
+        // Map the helper HashMap to an indexed vec.
+        // The index references are 1-index based, so we can
+        // safely prepend a "root" node at 0.
+        let mut nodes = Vec::with_capacity(self.nodes.len() + 1);
+        nodes.push((None, root_label, (outputs, inputs)));
+        for (label, _, node, io) in self.nodes.into_iter() {
+            nodes.push((Some(node), label, io));
+        }
+        PortGraph::new(nodes, self.edges)
+    }
+}
+
+impl<T> std::default::Default for GraphBuilder<T>
+where
+    T: Debug,
+{
+    fn default() -> Self {
+        GraphBuilder {
+            nodes: vec![],
+            edges: vec![],
+            inputs: None,
+            outputs: None,
+            root_label: None,
+        }
+    }
+}
+
+/// Genericize taking an iterator of a string-like.
+fn serialize_io<I, S>(def: I) -> Vec<String>
+where
+    I: IntoIterator<Item = S>,
+    S: Into<String>,
+{
+    def.into_iter().map(|s| s.into()).collect()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn s(s: &str) -> String {
+        s.into()
+    }
+
+    #[test]
+    fn it_builds_a_graph() -> Result<()> {
+        let builder = GraphBuilder::default();
+        let graph = builder
+            .set_label("RootNode")
```

Comment:
Not happy with this API, but also don't want to spend too much time on it before full integration. It handles defining nodes by keys, rather than raw indices like the underlying `PortGraph`